### PR TITLE
Refactor Classifier & Head

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

- Refactor Classifier & Head
- Refactor Torchvision Model (nn.Module & OTXModel)
- Add unit-test for classifier

image

- torchvision model
image

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/classification/efficientnet.py [hunk 1]
```diff
@@ -106,9 +104,8 @@ def _build_model(self, num_classes: int) -> nn.Module:
             head=LinearClsHead(
                 num_classes=num_classes,
                 in_channels=backbone.num_features,
-                topk=(1, 5) if num_classes >= 5 else (1,),
-                loss=loss,
             ),
+            loss=nn.CrossEntropyLoss(),
```
Comment (eunwoosh) [inline src/otx/algo/classification/efficientnet.py:111]: It's just a question. Default value of reuction of cross entropy loss is "mean" so, loss doesn't have to be divided by `cls_score.size(0)`. Am I understanding right?

File: src/otx/algo/classification/heads/multilabel_cls_head.py [hunk 2]
```diff
@@ -180,7 +125,6 @@ class MultiLabelLinearClsHead(MultiLabelClsHead):
         in_channels (int): Number of channels in the input feature map.
         normalized (bool): Normalize input features and weights.
         scale (float): positive scale parameter.
```
Comment (eunwoosh) [inline src/otx/algo/classification/heads/multilabel_cls_head.py]:
```suggestion
```

File: src/otx/algo/classification/heads/multilabel_cls_head.py [hunk 3]
```diff
@@ -61,7 +59,6 @@ class MultiLabelClsHead(BaseModule):
         scale (float): The scaling factor for the classification score.
```
Comment (eunwoosh) [inline src/otx/algo/classification/heads/multilabel_cls_head.py]: This docstring looks quite different than current model class. Could you align it to current implementation?

File: src/otx/algo/classification/heads/multilabel_cls_head.py [hunk 4]
```diff
@@ -244,7 +180,6 @@ class MultiLabelNonLinearClsHead(MultiLabelClsHead):
         activation_callable (Callable[..., nn.Module]): Activation layer module.
             Defaults to nn.ReLU.
         scale (float): Positive scale parameter.
```
Comment (eunwoosh) [inline src/otx/algo/classification/heads/multilabel_cls_head.py]:
```suggestion
```

File: src/otx/algo/classification/classifier/base_classifier.py [hunk 5]
```diff
@@ -27,30 +30,14 @@ class ImageClassifier(BaseModule):
...
 
     Args:
-        backbone (dict): The backbone module. See
-            :mod:`mmpretrain.models.backbones`.
-        neck (dict, optional): The neck module to process features from
-            backbone. See :mod:`mmpretrain.models.necks`. Defaults to None.
-        head (dict, optional): The head module to do prediction and calculate
-            loss from processed features. See :mod:`mmpretrain.models.heads`.
+        backbone (nn.Module): The backbone module.
+        neck (nn.Module, optional): The neck module to process features from
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/base_classifier.py]:
```suggestion
        neck (nn.Module | None, optional): The neck module to process features from
```

File: src/otx/algo/classification/classifier/base_classifier.py [hunk 6]
```diff
@@ -115,7 +98,7 @@ def forward(
         msg = f'Invalid mode "{mode}".'
         raise RuntimeError(msg)
 
-    def extract_feat(self, inputs: torch.Tensor, stage: str = "neck") -> tuple | torch.Tensor:
+    def extract_feat(self, inputs: torch.Tensor, stage: str = "neck") -> torch.Tensor:
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/base_classifier.py:100]: Could you update return type docstring also?

File: src/otx/algo/classification/classifier/h_label_classifier.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+from otx.algo.classification.heads.hlabel_cls_head import HierarchicalClsHead
+from otx.algo.classification.utils.ignored_labels import get_valid_label_mask
+
+from .base_classifier import ImageClassifier
+
+if TYPE_CHECKING:
+    from torch import nn
+
+
+class HLabelClassifier(ImageClassifier):
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/h_label_classifier.py:22]: Could you add docstring for the class?

File: src/otx/algo/classification/classifier/h_label_classifier.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    """HLabel Classifier."""
+
+    head: HierarchicalClsHead
+
+    def __init__(
+        self,
+        backbone: nn.Module,
+        neck: nn.Module | None,
+        head: nn.Module,
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/h_label_classifier.py]:
```suggestion
        head: HierarchicalClsHead,
```

File: src/otx/algo/classification/classifier/h_label_classifier.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        backbone_feat = x
+
+        feature_vector = feature_vector_fn(backbone_feat)
+        saliency_map = self.explainer.func(backbone_feat)
+
+        if hasattr(self, "neck") and self.neck is not None:
+            x = self.neck(x)
+
+        logits = self.head(x)
+        pred_results = self.head._get_predictions(logits)  # noqa: SLF001
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/h_label_classifier.py:131]: These two lines are only difference compared to `ImageClassifier`, right? How about making a function to get `logits`, `scores` and `preds` and using it in `_forward_explain` function of the parent class and override that function only? This is my idea. if you think differently, please do as you think :)

File: src/otx/algo/classification/classifier/semi_sl_classifier.py [hunk 10]
```diff
@@ -10,22 +10,45 @@
...
 
 import torch
+from torch import nn
+
+from otx.algo.classification.heads.semi_sl_head import OTXSemiSLClsHead
 
 from .base_classifier import ImageClassifier
 
 
 class SemiSLClassifier(ImageClassifier):
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/semi_sl_classifier.py:20]: Could you update argument doctstring?

File: src/otx/algo/classification/classifier/semi_sl_classifier.py [hunk 11]
```diff
@@ -10,22 +10,45 @@
...
+        )
+        self.unlabeled_coef = unlabeled_coef
+
     def extract_feat(
         self,
-        inputs: dict[str, torch.Tensor],
+        inputs: dict[str, torch.Tensor] | torch.Tensor,
         stage: str = "neck",
-    ) -> dict[str, tuple | torch.Tensor] | tuple | torch.Tensor:
+    ) -> dict[str, torch.Tensor] | tuple[torch.Tensor] | torch.Tensor:
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/semi_sl_classifier.py:56]: Could you update docstring for this?

File: src/otx/algo/classification/torchvision_model.py [hunk 12]
```diff
@@ -23,422 +30,128 @@
...
-            The type of classification task.
-        train_type (Literal[OTXTrainType.SUPERVISED, OTXTrainType.SEMI_SUPERVISED], optional): The type of training.
-        head_config (dict | None, optional): The configuration for the head module.
-
-    Methods:
-        forward(images: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
-            Performs forward pass of the model.
-
-    """
+class TVModelForMulticlassCls(OTXMulticlassClsModel):
```
Comment (eunwoosh) [inline src/otx/algo/classification/torchvision_model.py:50]: Could you add argument docstring?

File: src/otx/algo/classification/torchvision_model.py [hunk 13]
```diff
@@ -23,422 +30,128 @@
...
-        freeze_backbone (bool, optional): Whether to freeze the backbone model. Defaults to False.
-        task (Literal[OTXTaskType.MULTI_CLASS_CLS, OTXTaskType.MULTI_LABEL_CLS, OTXTaskType.H_LABEL_CLS], optional):
-            The type of classification task.
-        train_type (Literal[OTXTrainType.SUPERVISED, OTXTrainType.SEMI_SUPERVISED], optional): The type of training.
-        input_size (tuple[int, int], optional):
-            Model input size in the order of height and width. Defaults to (224, 224)
-    """
-
-    model: TVClassificationModel
+class TVModelForMultilabelCls(OTXMultilabelClsModel):
```
Comment (eunwoosh) [inline src/otx/algo/classification/torchvision_model.py:158]: Could you add argument docstring?

File: src/otx/algo/classification/torchvision_model.py [hunk 14]
```diff
@@ -451,219 +164,130 @@ def __init__(
...
-                "train/num_pseudo_label",
-                self.model.head.num_pseudo_label,
-                on_step=True,
-                on_epoch=False,
-                prog_bar=True,
-            )
-        return loss
 
-    def forward_explain(self, inputs: CLASSIFICATION_BATCH_DATA_ENTITY) -> CLASSIFICATION_BATCH_PRED_ENTITY:
+class TVModelForHLabelCls(OTXHlabelClsModel):
```
Comment (eunwoosh) [inline src/otx/algo/classification/torchvision_model.py:252]: Could you update arguemnt docstring?

File: src/otx/algo/classification/utils/ignored_labels.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+"""Util functions related to ignored_labels."""
+
+from __future__ import annotations
+
+import torch
+
+from otx.core.data.entity.base import ImageInfo
+
+
+def get_valid_label_mask(img_metas: list[ImageInfo], num_classes: int) -> torch.Tensor:
```
Comment (eunwoosh) [inline src/otx/algo/classification/utils/ignored_labels.py:13]: Could you add unit test for this class?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work :) I left some comments. Please take a look.

File: src/otx/algo/classification/efficientnet.py (hunk 1 above)
Comment (harimkang) [reply src/otx/algo/classification/efficientnet.py:111]: Yes, you are correct. Instead of dividing by `cls_score.size(0)` in the `loss function` like we did before, i changed it to use mean by default, so the behavior is the same.

File: src/otx/algo/classification/classifier/h_label_classifier.py (hunk 9 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/h_label_classifier.py:131]: Yeah I'll have to think about it and work on it later, I think it would be a head refactoring, thanks for the good idea.

File: src/otx/algo/classification/heads/multilabel_cls_head.py (hunk 2 above)
Comment (harimkang) [reply src/otx/algo/classification/heads/multilabel_cls_head.py]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/pull/3830/commits/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/heads/multilabel_cls_head.py (hunk 3 above)
Comment (harimkang) [reply src/otx/algo/classification/heads/multilabel_cls_head.py]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/pull/3830/commits/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/heads/multilabel_cls_head.py (hunk 4 above)
Comment (harimkang) [reply src/otx/algo/classification/heads/multilabel_cls_head.py]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/pull/3830/commits/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/classifier/base_classifier.py (hunk 5 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/base_classifier.py]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/pull/3830/commits/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/classifier/base_classifier.py (hunk 6 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/base_classifier.py:100]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/classifier/h_label_classifier.py (hunk 7 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/h_label_classifier.py:22]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/classifier/h_label_classifier.py (hunk 8 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/h_label_classifier.py]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/classifier/semi_sl_classifier.py (hunk 10 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/semi_sl_classifier.py:20]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/classifier/semi_sl_classifier.py (hunk 11 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/semi_sl_classifier.py:56]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/torchvision_model.py (hunk 12 above)
Comment (harimkang) [reply src/otx/algo/classification/torchvision_model.py:50]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/torchvision_model.py (hunk 13 above)
Comment (harimkang) [reply src/otx/algo/classification/torchvision_model.py:158]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/torchvision_model.py (hunk 14 above)
Comment (harimkang) [reply src/otx/algo/classification/torchvision_model.py:252]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/utils/ignored_labels.py (hunk 15 above)
Comment (harimkang) [reply src/otx/algo/classification/utils/ignored_labels.py:13]: Thanks! Done - https://github.com/openvinotoolkit/training_extensions/commit/5ed3863c2f7d3db9b94acbbf9530abbce28536ba

File: src/otx/algo/classification/classifier/h_label_classifier.py (hunk 8 above)
Comment (eunwoosh) [reply src/otx/algo/classification/classifier/h_label_classifier.py]: Thanks! could you remove 43line?

File: src/otx/algo/classification/heads/multilabel_cls_head.py (hunk 3 above)
Comment (eunwoosh) [reply src/otx/algo/classification/heads/multilabel_cls_head.py]: It looks different still. Could you check again?

Comment (harimkang) [reply src/otx/algo/classification/heads/multilabel_cls_head.py]: you meant here? - https://github.com/openvinotoolkit/training_extensions/pull/3830/commits/da3a08ca79e33510e3904dbd2a71ff1f53f4a6c8

File: src/otx/algo/classification/classifier/h_label_classifier.py (hunk 8 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/h_label_classifier.py]: https://github.com/openvinotoolkit/training_extensions/pull/3830/commits/da96f40db10fc64fa800b6b3aadcec9c49b75131

**Review by eunwoosh — DISMISSED** [review]

**Review by sungchul1 — DISMISSED** [review]

**Review by sungchul1 — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]
